File: Algo1/PA 1/merge.py
import time;
inversions = 0

def countInversions(array, length):
    global inversions
    for i in range(length):
        for j in range(i + 1, length):
            if array[i] > array[j]:
                inversions += 1

def mergeSort(array):
    global inversions

    if len(array) < 2:
        # print(f'Base case return: {array}')
        return array

    # countInversions(array, len(array))
    A = mergeSort(array[:int(len(array)/2)])
    B = mergeSort(array[int(len(array)/2):])
    length = len(A)
    i = 0;
    j = 0;

    # print(f'Current array: {array}')
    # print(f'A array: {A}')
    # print(f'B array: {B}')
    for k in range(len(array)):
        if not A:

            array[k] = B[j]
            if j + 1 == len(B):
                B = []
            else:
                j += 1

        elif not B:

            array[k] = A[i]
            if i + 1 == len(A):
                A = []
            else:
                i += 1

        elif A and A[i] < B[j]:

            array[k] = A[i]
            if i + 1 == len(A):
                A = []
            else:
                i += 1

        elif B and B[j] < A[i]:
            # print(f'Array A: {A[i:]}|{A}\nArray B: {B[j:]}|{B}')
            # with open('result.txt', 'a') as f:
            #     f.write(f'\nArray A: {A[i:]}|{A}\nArray B: {B[j:]}|{B}\n')

            inversions += length - i
            array[k] = B[j]
            if j + 1 == len(B):
                B = []
            else:
                j += 1

        elif B[j] == A[i]:

            array[k] = B[j]
            if j + 1 == len(B):
                B = []
            else:
                j += 1


    return array

def bubbleSort(array):
    temp = 0
    i = len(array)
    while i:
        for j in range(len(array) - 1):
            if array[j] > array[j + 1]:
                temp = array[j]
                array[j] = array[j + 1]
                array[j + 1] = temp
        i -= 1
    return array


with open('IntegerArray.txt', 'r') as f:
    nums = [int(item.replace('\n', '')) for item in f.readlines()]

with open('result.txt', 'w') as f:
    f.write('')
start = time.time();
# nums.reverse()
nums = mergeSort(nums)
# with open('Output.txt', 'w') as f:
#     for num in nums:
#         f.write(f"{num}\n")
# print(nums)
print(f'Merge sort executed in {time.time() - start} seconds')
print(f'Total inversions: {inversions}')

# start = time.time();
# bubbleSort(nums)
# print(f'Bubble sort executed in {time.time() - start} seconds')

File: Algo1/PA 1/merge.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

int * copy(int * input, int length){
	int * copy = (int*)malloc(length * sizeof(int));
	int * start = copy;


	*copy = *input;
	copy++, input++;

	while(length > 1){
		//printf("Length: %d", length);
		*copy = *input;
		copy++, input++, length--;
	}

	return start;
}

int * mergeSort(int * array, int n, double * inversions){
	int * temp = array;
	int * A;
	int * B;
	int * startA;
	int * startB;
	int lenB = 0;
	int lenA = 0;
	int i = 0;
	int j = 0;
	//printf("\n\n-----------------------\nFirst element of current array: %d\nLength: %d\n-----------------------\n\n", *array, n);

	if(n < 2){
		//printf("\nReturned array\n");
		return array;
	}

	else if(n%2 == 1){
		//printf("\n\n-----------------------\nFirst element of current array: %d\nLength: %d\n-----------------------\n\n", *array, n);
		lenA = (n - 1)/2;
		lenB = (n - 1)/2 + 1;
		A = mergeSort(copy(array, lenA), lenA, inversions);
		startA = A;
		B = mergeSort(copy(array + lenA, lenB), lenB, inversions);
		startB = B;
	}

	else{
		//printf("\n\n-----------------------\nFirst element of current array: %d\nLength: %d\n-----------------------\n\n", *array, n);
		lenA = n/2;
		lenB = n/2;
		A = mergeSort(copy(array, lenA), lenA, inversions);
		startA = A;
		B = mergeSort(copy(array + lenB, lenB), lenB, inversions);
		startB = B;
	}

	for(int k = 0; k < n; k++){
		if(A == NULL){

			*temp = *B;
			if(j + 1 == lenB) B = NULL;
			else j++, B++;
		}

		else if(B == NULL){

			*temp = *A;
			if(i + 1 == lenA) A = NULL;
			else i++, A++;
		}

		else if(A != NULL && *A < *B){
			*temp = *A;
			if(i + 1 == lenA) A = NULL;
			else i++, A++;

		}

		else if(B != NULL && *B < *A){
			*inversions += lenA - i;
			*temp = *B;
			if(j + 1 == lenB) B = NULL;
			else j++, B++;

		}

		else if(*B == *A){
			*temp = *B;
			if(j + 1 == lenB) B = NULL;
			else j++, B++;
		}

		temp++;
	}

	free(startA);
	free(startB);
	return array;
}
int main(){
	FILE * fp;
	char buff[255];
	int array[100000];
	int length = sizeof(array)/sizeof(int);
	int i = 0;
	double * inversions = (double*)malloc(sizeof(double));
	*inversions = 0;

	clock_t start, end;
	double cpu_time_used;

	fp = fopen("IntegerArray.txt", "r");
	while(fscanf(fp, "%s", buff) == 1){
		int value = atoi(buff);
		//printf("%d\n", value);
		array[i] = value;
		i++;

	}
	fclose(fp);
	start = clock();
	int * result = mergeSort(array, length, inversions);
	free(inversions);
	end = clock();
	cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
	printf("Merge Sort completed in %f seconds with %.f inversions\n", cpu_time_used, *inversions);

	//printf("\n");
	//for(i = 0; i < length; i++){
	//	printf("%d, ", result[i]);
	//}
	//printf("\n");

}
